翻译v3.4原生反射机制 (#1942)
cocos-creator/creator-docs: en/advanced-topics/js-oc-bridge.md

# A Simpler Way to Call Objective-C Methods with JavaScript (Experimental)

## Background

> Prior to v3.4.0, the reflection mechanism in [Using JavaScript to Call Objective-C](./oc-reflection.md) static methods, we not only needed to strictly declare package names and function signatures, but also needed to strictly check the number of parameters to ensure proper operation, which was a complicated step.

Reviewer: ```suggestion
Prior to v3.4.0, the reflection mechanism in [Using JavaScript to Call Objective-C](./oc-reflection.md) static methods, not only needed to strictly declare package names and function signatures, but also needed to strictly check the number of parameters to ensure proper operation, which was a complicated step.
```

> So in v3.4.0 we additionally provide another experimental method for simplifying calls from the scripting layer to the native layer. This is a kind of channel, or a bridge, which we named `JsbBridge` before introducing other scripting systems, meaning that it serves as a bridge to communicate between script and native APP via the `JSB` binding.

Reviewer: ```suggestion
Additionally provided in v3.4.0 is another experimental method for simplifying calls from the scripting layer to the native layer. This is a kind of channel, or a bridge, named `JsbBridge` before introducing other scripting systems, meaning that it serves as a bridge to communicate between script and native APP via the `JSB` binding.
```

> > **Note**: both ways are working fine, developers can choose to use them according to their actual needs. To use the previous way, please review the  [Using JavaScript to Call Objective-C](./oc-reflection.md).

Reviewer 2: ```suggestion
> **Note**: both ways are working fine, developers can choose to use them according to their actual needs. To use the previous way, please review the  [Using JavaScript to Call Objective-C](./oc-reflection.md) documentation.
```

> > **Note**: both ways are working fine, developers can choose to use them according to their actual needs. To use the previous way, please go to [Using JavaScript to Call Objective-C](./oc-reflection.md) documentation to see.

Reviewer: ```suggestion
> **Note**: both ways are working fine, developers can choose to use them according to their actual needs. To use the previous way, please review the  [Using JavaScript to Call Objective-C](./oc-reflection.md).
```

## JavaScript Interface Introduction

The only two interfaces at the scripting level are `sendToNative` and `onNative`, which are **transfer** and **receive native layer** parameters, respectively. The following points need to be noted when using them:

> - Since this feature is still in the experimental stage, only `string` transfers are supported. If you need to transfer objects containing multiple parameters, please consider converting them to `Json` form for transfer and parsing them at different levels.

Reviewer: ```suggestion
- This feature is still in the experimental stage, only `string` transfers are supported. To transfer objects containing multiple parameters, please consider converting them to the `JSON` format for transfer and parsing them at different levels.
```

@end
```

## Basic Usage

> ### Using JavaScript to trigger Objective-C Callbacks

Reviewer: ```suggestion
### Using JavaScript to Trigger Objective-C Callbacks
```

> Assuming our ad interface is set in the native layer, then when the player clicks the button to open the ad, it is logical to trigger `ObjC` to open the ad.

Reviewer: ```suggestion
Assuming the ad interface is set in the native layer, then when the player clicks the button to open the ad, it is logical to trigger `ObjC` to open the ad.
```

> The code example for opening th ad interface is as follows:

Reviewer: ```suggestion
The code example for opening the ad interface is as follows:
```

> The code example for opening the ad interface is as follows:

Reviewer 2: ```suggestion
The code example of the interface to open the ad is as follows:
```

> The code example for opening the ad interface is as follows:

Reviewer 2: ```suggestion
The code example of the interface to open the ad is as follows:
```

```ObjC
static ICallback cb = ^void (NSString* _arg0, MSString* _arg1){
    //open Ad
}
```

> At this point, register the event to open the ad first:

Author: ```suggestion
At this point, register the event that opens the ad first:
```

> At this point we need to register the event to open the ad first:

Reviewer: ```suggestion
At this point, register the event to open the ad first:
```

```ObjC
JsbBridge* m = [JsbBridge sharedInstance];
[m setCallback:cb];
```

> Perform the open action on the button click event in the `JavaScript` layer script:

Reviewer 2: ```suggestion
Perform the open action on the button's click event in the `JavaScript` layer script:
```

> And perform the open action on the button click event in the `JavaScript` layer script:

Reviewer: ```suggestion
Perform the open action on the button click event in the `JavaScript` layer script:
```

} 
```

This will send the required information to the `ObjC` layer through the `Jsb.Bridge` channel to perform the action of opening the ad.

> ### Using Objective-C to trigger JavaScript

Reviewer 2: ```suggestion
### Using Objective-C to trigger JavaScript Callbacks
```

> Assuming that our animation playback operation is recorded in JavaScript and we want to play this animation in the Objective-C layer, we can also register an event to play the animation.

Reviewer: ```suggestion
Assume that the animation playback operation is recorded in JavaScript. To play this animation in the Objective-C layer, register an event to play the animation.
```

> First you need to define a function to play the animation:

Reviewer: ```suggestion
First, define a function to play the animation:
```

```ts
public void playAnimation(animationName: string, isLoop: boolean){

>     //Code to play Animation

Reviewer 2: ```suggestion
    // Code to play Animation
```

}
```

> Then document the method in `onNative`:

Reviewer: ```suggestion
TNext, document the method in `onNative`:
```

> Then document the method in `onNative`:

Author: ```suggestion
Next, document the method in `onNative`:
```

}
```

The `ObjC` code example is as follows:

```ObjC
JsbBridge* m = [JsbBridge sharedInstance];
[m sendToScript:@"Animation1" arg1:@"true"];
```

This will call the `JavaScript` playback operation.

## Sample project: simple multi-event calls

> The Creator provides the [native-script-bridge](https://github.com/cocos-creator/example-3d/tree/v3.4/native-script-bridge) example, which developers can download for reference use as needed.

Reviewer: ```suggestion
Creator provides the [native-script-bridge](https://github.com/cocos-creator/example-3d/tree/v3.4/native-script-bridge) example, which developers can download for reference use as needed.
```
